projects/packages/forms/src/blocks/field-name/hooks/use-name-field-transforms.js

```diff
@@ -0,0 +1,124 @@
+ * @param {string} params.id       - Current field id (used to infer variant).
+ */
+export default function useNameFieldTransforms( { clientId, id } ) {
+	const prevIdRef = useRef( id );
+	const { updateBlockAttributes } = useDispatch( blockEditorStore );
+
+	const existingFieldIds = useSelect(
+		select => {
+			const { getBlock, getBlocks, getBlockParents } = select( blockEditorStore );
+			const parentIds = getBlockParents( clientId ) || [];
+			const formRootId = parentIds.find( parentId => {
+				const parent = getBlock( parentId );
+				return parent?.name === 'jetpack/contact-form';
+			} );
+			const siblings = formRootId ? getBlocks( formRootId ) : getBlocks();
+			return siblings
+				.filter( block => block.clientId !== clientId && block.attributes?.id )
+				.map( block => block.attributes.id );
+		},
+		[ clientId ]
+	);
+
+	const labelClientId = useSelect(
+		select => {
+			const block = select( blockEditorStore ).getBlock( clientId );
+			const labelBlock = block?.innerBlocks?.find( b => b.name === 'jetpack/label' );
+			return labelBlock?.clientId;
+		},
+		[ clientId ]
+	);
+
+	// Derive block context details once per ID change.
+	const context = useMemo( () => {
+		const newId = id;
+		const prevId = prevIdRef.current;
+		const isTransformingToBaseBlock =
+			isKnownNameVariationId( prevId ) && ( newId === undefined || newId === '' );
+		const isTransformingToVariation = isKnownNameVariationId( newId ) && newId !== prevId;
+
+		let baseId = NAME_ID;
+		if ( isFirstNameVariationId( newId ) ) {
+			baseId = FIRST_NAME_ID;
+		} else if ( isLastNameVariationId( newId ) ) {
+			baseId = LAST_NAME_ID;
+		}
+
+		return {
+			isTransformingToBaseBlock,
+			isTransformingToVariation,
+			baseId,
+			label: getDefaultLabelForId( baseId ),
+		};
+	}, [ id ] );
+
+	// Set HTML ID on the name field block.
+	useEffect( () => {
+		if ( context.isTransformingToBaseBlock ) {
+			// Only clear the id if it isn't already empty to avoid loops.
+			if ( id !== '' ) {
+				updateBlockAttributes( clientId, { id: '' } );
+			}
+			return;
+		}
+		if ( context.isTransformingToVariation ) {
+			const uniqueId = generateUniqueFormFieldId( context.baseId, existingFieldIds );
+			// Only set the id when it actually changes to avoid loops.
+			if ( id !== uniqueId ) {
+				updateBlockAttributes( clientId, { id: uniqueId } );
+			}
+		}
+	}, [ context, id, clientId, updateBlockAttributes, existingFieldIds ] );
+
+	// Set label on the label block.
+	useEffect( () => {
+		if ( ! labelClientId ) {
+			return;
+		}
+		if ( context.isTransformingToBaseBlock ) {
+			updateBlockAttributes( labelClientId, { label: DEFAULT_NAME_LABEL } );
+			return;
+		}
+		if ( context.isTransformingToVariation ) {
+			updateBlockAttributes( labelClientId, { label: context.label } );
+		}
+	}, [ context, labelClientId, updateBlockAttributes ] );
+
+	// Track previous ID.
+	useEffect( () => {
+		prevIdRef.current = id;
+	}, [ id ] );
+}
```

Comment:
We had a useNameLabelSync hook. This one replace it. It does the following: 
* Get's existingFieldIds (so we can avoid duplication) and labelClientId (so we can updated the label
* Generates a contact object/variable that contains things we need to know to update the ID and label
* Adds useEffects to update the ID and label, and track ID for transforms